[Operator Versioning][Edge] Make to_tuple a shared function

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67731
* #67728
* #67730
* #67729
* __->__ #67727
* #67726



`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852)

Commit:
[Operator Versioning][Edge] Make to_tuple a shared function

`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -38,9 +38,7 @@
-IValue to_tuple(std::vector<IValue> ivalues) {
-  return c10::ivalue::Tuple::create(std::move(ivalues));
-}
+char const* toString(OpCode op);
```
@zhxchen17 I remember you also made this change before? Has it been landed?

zhxchen17 replied:
@tugsbayasgalan yeah I did, but then I realized that's not necessary for my diff so I didn't land that in the final version.

Commit:
Update on "[Operator Versioning][Edge] Make to_tuple a shared function"


`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Make to_tuple a shared function"


`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

[ghstack-poisoned]

Commit:
Update on "[Operator Versioning][Edge] Make to_tuple a shared function"




`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852)

[ghstack-poisoned]

iseeyuan commented:
Since we don't use deserialization, is to_tuple still needed?

Commit:
Update on "[Operator Versioning][Edge] Make to_tuple a shared function"




`to_tuple` is a common function and is used in `upgrader_mobile.h`. Make it sharable between import and export.

Example upgrader_mobile.h
```
const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
    {{std::string("aten::div_Tensor"),
      std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})}});

static const std::vector<std::pair<std::string, c10::IValue>> kUpgraderBytecode(
    {
        // The following are called from setup sections.
        {"div_Tensor_0_3",
         to_tuple(std::vector<c10::IValue>{
             //  instructions
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"STOREN", 1, 2}), to_tuple({"LOAD", 1, 0}),
                 to_tuple({"OP", 0, 0}),     to_tuple({"JF", 3, 0}),
                 to_tuple({"LOADC", 1, 0}),  to_tuple({"JMP", 3, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 0, 0}),
                 to_tuple({"STORE", 3, 0}),  to_tuple({"MOVE", 3, 0}),
                 to_tuple({"JF", 5, 0}),     to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"OP", 1, 0}),
                 to_tuple({"JMP", 5, 0}),    to_tuple({"LOAD", 1, 0}),
                 to_tuple({"LOAD", 2, 0}),   to_tuple({"LOADC", 0, 0}),
                 to_tuple({"OP", 2, 0}),     to_tuple({"STORE", 4, 0}),
                 to_tuple({"DROPR", 2, 0}),  to_tuple({"DROPR", 1, 0}),
                 to_tuple({"MOVE", 4, 0}),   to_tuple({"RET", 0, 0}),
             }),
             //  operators
             to_tuple(std::vector<c10::IValue>{
                 to_tuple({"aten::is_floating_point", "", 1}),
                 to_tuple({"aten::div", "Tensor", 2}),
                 to_tuple({"aten::div", "Tensor_mode", 3}),
             }),
             //  constants
             to_tuple(std::vector<c10::IValue>{
                 c10::IValue("trunc"),
                 c10::IValue(true),
             }),
             //  types
             to_tuple(std::vector<c10::IValue>{

             }),
             //  register_size
             to_tuple(std::vector<c10::IValue>{
                 4,
             }),
         })},
    });
```

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852/)

Differential Revision: [D32086852](https://our.internmc.facebook.com/intern/diff/D32086852)

[ghstack-poisoned]

cccclai commented:
> Since we don't use deserialization, is to_tuple still needed?

Good point. It's probably not needed.

iseeyuan commented:
If it's confirmed not needed, please remember to remove it from the stack.
